### ISSUE TITLE
Generate iPXE chain from autodetect

### ISSUE THREAD

**Issue by sjmh** [cobbler/cobbler#1641]
One of the issues with the current setup in cobbler when using iPXE is that you *have* to run a sync after each new system is put into cobbler.  This is because the iPXE configs are generated via a custom URL that includes the system name.

iPXE, however, knows the MAC address of the host that is booting.  It can be configured instead to send a request back to the cobbler server, which could then provide it an iPXE script with the hostname that matches the MAC.  The 'real' gPXE url could then be used to get to the original {g,i}PXE config.

So, for instance, all clients could be configured to boot into iPXE and retrieve this config:

```
#!ipxe

# get the MAC address
set dhcp_mac ${net0/mac:hex}
set cobbler ${next-server}

# this will set a var called 'sysname'
chain --autofree http://${cobbler}/cblr/svc/op/ipxedetect/mac/${dhcp_mac} ||

# chain to our real gpxe config
chain http://${cobbler}/cblr/svc/op/gpxe/system/${sysname}
```

The 'op/ipxedetect' function returns:

```
#!ipxe
set sysname <name of system>
```

And the dhcpd.conf could be configured like so:

```
# ******************************************************************
# Cobbler managed dhcpd.conf file
# generated from cobbler dhcp.conf template (Wed Mar 16 22:04:40 2016)
# Do NOT make changes to /etc/dhcpd.conf. Instead, make your changes
# in /etc/cobbler/dhcp.template, as /etc/dhcpd.conf will be
# overwritten.
# ******************************************************************

ddns-update-style interim;

allow booting;
allow bootp;

ignore client-updates;
set vendorclass = option vendor-class-identifier;

option pxe-system-type code 93 = unsigned integer 16;

subnet 10.6.7.0 netmask 255.255.255.0 {
    option routers             10.6.7.1;
    option domain-name-servers 10.6.7.20;
    option subnet-mask         255.255.255.0;
    range dynamic-bootp        10.6.7.100 10.6.7.254;
    default-lease-time         21600;
    max-lease-time             43200;
    next-server                10.6.7.20;
    if exists user-class and option user-class = "gPXE" {
        filename "http://10.6.7.20:80/cobbler/pub/boot.ipxe";
    } else if exists user-class and option user-class = "iPXE" {
        filename "http://10.6.7.20:80/cobbler/pub/boot.ipxe";
    } else {
        filename "undionly.kpxe";
    }
}
```

And now, a sync is never required for iPXE clients.

The only real modification to cobbler here is the need for an autodetect function that takes the MAC address as an argument ( since iPXE can't send it as a header ) and returns the small iPXE script to set the sysname variable.

This could either be included as it's own function ( ipxedetect ), or autodetect could be modified to accept something like:

http://cobblerserv/cblr/svc/op/autodetect/ipxe/<mac address>

Thoughts?

**Comment by sjmh** [comment]
Actually, it may just be easier to write an 'ipxe' function for services.py that returns the rendered ipxe config based on the mac address.  I did it the other way because I didn't want to spend the time trying to figure out how to render the ipxe file.

**Comment by sjmh** [comment]
And nevermind! Looks like this is already implemented in master.  This would be nice to get backported to 28 and 26 though.

**Comment by Ankitmaroo** [comment]
Hi, I am trying to apply this feature in my Cobbler 2.6.11 release install and was wondering how is it already implemented in master branch?  Could you please help me understand on how to use it from master code?

**Comment by sjmh** [comment]
We embed the following into our undionly.kpxe:

```
#!ipxe

echo Trying DHCP
:retry_boot
ifconf -c dhcp || goto retry_boot

# chain to our real gpxe config
echo Getting ipxe config for ${netX/mac}
chain http://${next-server}/cblr/svc/op/gpxe/mac/${netX/mac}
```

**Comment by Ankitmaroo** [comment]
Sorry if these questions are lame:
The file https://github.com/cobbler/cobbler/blob/master/bin/services.py doesnt seem to have this function to support it

**Comment by sjmh** [comment]
Nope, it's only in the latest.  You can just copy the version from the master branch and put it into your services.py

**Comment by Ankitmaroo** [comment]
I replaced services in python package, in web folder, deleted pyc files and rebooted the server but it still doesnt pick up the function .... Are there any changes I am missing. I would not want to reinstall the entire thing.
